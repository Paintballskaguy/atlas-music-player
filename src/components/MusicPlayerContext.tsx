import React, { createContext, useContext, useReducer, useEffect, ReactNode } from 'react';
import ApiService from '../services/api';
import { PlaylistSong, Song, MusicPlayerState, MusicPlayerContextType, RepeatMode } from '../types';

const MusicPlayerContext = createContext<MusicPlayerContextType | undefined>(undefined);

// Action types
const ACTIONS = {
  SET_LOADING: 'SET_LOADING',
  SET_PLAYLIST: 'SET_PLAYLIST',
  SET_CURRENT_SONG: 'SET_CURRENT_SONG',
  SET_CURRENT_SONG_DETAILS: 'SET_CURRENT_SONG_DETAILS',
  SET_LYRICS: 'SET_LYRICS',
  SET_PLAYING: 'SET_PLAYING',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_DURATION: 'SET_DURATION',
  SET_VOLUME: 'SET_VOLUME',
  SET_PLAYBACK_SPEED: 'SET_PLAYBACK_SPEED',
  SET_SHUFFLE: 'SET_SHUFFLE',
  SET_REPEAT: 'SET_REPEAT',
  SET_ERROR: 'SET_ERROR',
  CLEAR_ERROR: 'CLEAR_ERROR'
} as const;

type ActionType = typeof ACTIONS[keyof typeof ACTIONS];

type Action = 
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_PLAYLIST'; payload: PlaylistSong[] }
  | { type: 'SET_CURRENT_SONG'; payload: PlaylistSong | null }
  | { type: 'SET_CURRENT_SONG_DETAILS'; payload: Song | null }
  | { type: 'SET_LYRICS'; payload: string | null }
  | { type: 'SET_PLAYING'; payload: boolean }
  | { type: 'SET_CURRENT_TIME'; payload: number }
  | { type: 'SET_DURATION'; payload: number }
  | { type: 'SET_VOLUME'; payload: number }
  | { type: 'SET_PLAYBACK_SPEED'; payload: number }
  | { type: 'SET_SHUFFLE'; payload: boolean }
  | { type: 'SET_REPEAT'; payload: RepeatMode }
  | { type: 'SET_ERROR'; payload: string | null }
  | { type: 'CLEAR_ERROR' };

// Initial state
const initialState: MusicPlayerState = {
  playlist: [],
  currentSong: null,
  currentSongDetails: null,
  lyrics: null,
  isPlaying: false,
  currentTime: 0,
  duration: 0,
  volume: 50,
  playbackSpeed: 1,
  shuffle: false,
  repeat: false,
  isLoading: true,
  error: null
};

// Reducer
function musicPlayerReducer(state: MusicPlayerState, action: Action): MusicPlayerState {
  switch (action.type) {
    case ACTIONS.SET_LOADING:
      return { ...state, isLoading: action.payload };
    case ACTIONS.SET_PLAYLIST:
      return { ...state, playlist: action.payload };
    case ACTIONS.SET_CURRENT_SONG:
      return { ...state, currentSong: action.payload };
    case ACTIONS.SET_CURRENT_SONG_DETAILS:
      return { ...state, currentSongDetails: action.payload };
    case ACTIONS.SET_LYRICS:
      return { ...state, lyrics: action.payload };
    case ACTIONS.SET_PLAYING:
      return { ...state, isPlaying: action.payload };
    case ACTIONS.SET_CURRENT_TIME:
      return { ...state, currentTime: action.payload };
    case ACTIONS.SET_DURATION:
      return { ...state, duration: action.payload };
    case ACTIONS.SET_VOLUME:
      return { ...state, volume: action.payload };
    case ACTIONS.SET_PLAYBACK_SPEED:
      return { ...state, playbackSpeed: action.payload };
    case ACTIONS.SET_SHUFFLE:
      return { ...state, shuffle: action.payload };
    case ACTIONS.SET_REPEAT:
      return { ...state, repeat: action.payload };
    case ACTIONS.SET_ERROR:
      return { ...state, error: action.payload, isLoading: false };
    case ACTIONS.CLEAR_ERROR:
      return { ...state, error: null };
    default:
      return state;
  }
}

interface MusicPlayerProviderProps {
  children: ReactNode;
}

// Provider component
export const MusicPlayerProvider: React.FC<MusicPlayerProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(musicPlayerReducer, initialState);

  // Load initial playlist
  useEffect(() => {
    const loadPlaylist = async () => {
      try {
        dispatch({ type: ACTIONS.SET_LOADING, payload: true });
        const playlist = await ApiService.fetchPlaylist();
        dispatch({ type: ACTIONS.SET_PLAYLIST, payload: playlist });
        
        // Set first song as current if playlist exists
        if (playlist && playlist.length > 0) {
          dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: playlist[0] });
          await loadSongDetails(playlist[0].id);
        }
        
        dispatch({ type: ACTIONS.SET_LOADING, payload: false });
      } catch (error) {
        dispatch({ 
          type: ACTIONS.SET_ERROR, 
          payload: error instanceof Error ? error.message : 'Unknown error occurred'
        });
      }
    };

    loadPlaylist();
  }, []);

  // Actions
  const loadSongDetails = async (songId: string): Promise<void> => {
    try {
      const songDetails = await ApiService.fetchSong(songId);
      dispatch({ type: ACTIONS.SET_CURRENT_SONG_DETAILS, payload: songDetails });
    } catch (error) {
      console.error('Failed to load song details:', error);
    }
  };

  const loadLyrics = async (songId: string): Promise<void> => {
    try {
      const lyricsData = await ApiService.fetchLyrics(songId);
      dispatch({ type: ACTIONS.SET_LYRICS, payload: lyricsData.lyrics });
    } catch (error) {
      console.error('Failed to load lyrics:', error);
    }
  };

  const selectSong = async (song: PlaylistSong): Promise<void> => {
    dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: song });
    dispatch({ type: ACTIONS.SET_PLAYING, payload: false });
    dispatch({ type: ACTIONS.SET_CURRENT_TIME, payload: 0 });
    await loadSongDetails(song.id);
    await loadLyrics(song.id);
  };

  const togglePlayPause = (): void => {
    dispatch({ type: ACTIONS.SET_PLAYING, payload: !state.isPlaying });
  };

  const getNextSong = (): PlaylistSong | null => {
    if (!state.currentSong || !state.playlist.length) return null;

    if (state.shuffle) {
      const availableSongs = state.playlist.filter(song => song.id !== state.currentSong!.id);
      if (availableSongs.length === 0) return null;
      return availableSongs[Math.floor(Math.random() * availableSongs.length)];
    }

    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong!.id);
    if (currentIndex < state.playlist.length - 1) {
      return state.playlist[currentIndex + 1];
    }
    return null;
  };

  const nextSong = (): void => {
    const nextSong = getNextSong();
    if (nextSong) {
      selectSong(nextSong);
    }
  };

  const previousSong = (): void => {
    if (!state.currentSong) return;
    
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong!.id);
    if (currentIndex > 0) {
      const prevSong = state.playlist[currentIndex - 1];
      selectSong(prevSong);
    }
  };

  const setVolume = (volume: number): void => {
    dispatch({ type: ACTIONS.SET_VOLUME, payload: Math.max(0, Math.min(100, volume)) });
  };

  const setPlaybackSpeed = (speed: number): void => {
    dispatch({ type: ACTIONS.SET_PLAYBACK_SPEED, payload: speed });
  };

  const toggleShuffle = (): void => {
    dispatch({ type: ACTIONS.SET_SHUFFLE, payload: !state.shuffle });
  };

  const toggleRepeat = (): void => {
    const repeatModes: RepeatMode[] = [false, 'one', 'all'];
    const currentIndex = repeatModes.indexOf(state.repeat);
    const nextRepeat = repeatModes[(currentIndex + 1) % repeatModes.length];
    dispatch({ type: ACTIONS.SET_REPEAT, payload: nextRepeat });
  };

  const clearError = (): void => {
    dispatch({ type: ACTIONS.CLEAR_ERROR });
  };

  const contextValue: MusicPlayerContextType = {
    ...state,
    selectSong,
    togglePlayPause,
    nextSong,
    previousSong,
    setVolume,
    setPlaybackSpeed,
    toggleShuffle,
    toggleRepeat,
    loadLyrics,
    clearError
  };

  return (
    <MusicPlayerContext.Provider value={contextValue}>
      {children}
    </MusicPlayerContext.Provider>
  );
};

// Custom hook
export const useMusicPlayer = (): MusicPlayerContextType => {
  const context = useContext(MusicPlayerContext);
  if (!context) {
    throw new Error('useMusicPlayer must be used within a MusicPlayerProvider');
  }
  return context;
};