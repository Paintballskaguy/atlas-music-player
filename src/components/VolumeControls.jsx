import React, { useState } from 'react';
import { Volume2 } from 'lucide-react';

const VolumeControls = () => {
  const [volume, setVolume] = useState(50);

  const sliderStyle = {
    background: `linear-gradient(to right, #14b8a6 0%, #14b8a6 ${volume}%, #e5e7eb ${volume}%, #e5e7eb 100%)`
  };

  const darkSliderStyle = {
    background: `linear-gradient(to right, #5eead4 0%, #5eead4 ${volume}%, #4b5563 ${volume}%, #4b5563 100%)`
  };

  return (
    <>
      <div className="flex items-center space-x-4 w-full">
        <div className="p-2 rounded-full bg-custom-yellow-100 dark:bg-custom-yellow-800 transition-colors duration-200">
          <Volume2 
            size={24} 
            className="text-custom-yellow-700 dark:text-custom-yellow-200" 
          />
        </div>
        <div className="flex-1 relative">
          <input 
            type="range" 
            min="0" 
            max="100" 
            value={volume}
            onChange={(e) => setVolume(e.target.value)}
            className="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 dark:hidden"
            style={sliderStyle}
          />
          <input 
            type="range" 
            min="0" 
            max="100" 
            value={volume}
            onChange={(e) => setVolume(e.target.value)}
            className="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 hidden dark:block"
            style={darkSliderStyle}
          />
        </div>
        <span className="text-sm font-bold text-custom-teal-600 dark:text-custom-teal-300 min-w-[3rem] text-right">
          {volume}%
        </span>
      </div>
      
      <style dangerouslySetInnerHTML={{
        __html: `
          .volume-slider::-webkit-slider-thumb {
            appearance: none;
            width: 22px;
            height: 22px;
            border-radius: 50%;
            background: #dc2626;
            cursor: pointer;
            border: 3px solid white;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
            transition: all 0.2s ease;
          }
          
          .volume-slider::-webkit-slider-thumb:hover {
            background: #b91c1c;
            transform: scale(1.1);
            box-shadow: 0 6px 12px rgba(0, 0, 0, 0.4);
          }
          
          .volume-slider::-moz-range-thumb {
            width: 22px;
            height: 22px;
            border-radius: 50%;
            background: #dc2626;
            cursor: pointer;
            border: 3px solid white;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
            transition: all 0.2s ease;
          }
          
          .volume-slider::-moz-range-thumb:hover {
            background: #b91c1c;
            transform: scale(1.1);
          }
        `
      }} />
    </>
  );
};

export default VolumeControls;