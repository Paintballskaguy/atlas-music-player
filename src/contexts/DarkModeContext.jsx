import React, { createContext, useContext, useEffect, useState } from 'react';

const DarkModeContext = createContext();

export const useDarkMode = () => {
  const context = useContext(DarkModeContext);
  if (!context) {
    throw new Error('useDarkMode must be used within a DarkModeProvider');
  }
  return context;
};

// Export the provider as a named export
export const DarkModeProvider = ({ children }) => {
  const [isDark, setIsDark] = useState(() => {
    // Initialize from localStorage or system preference
    const saved = localStorage.getItem('darkMode');
    if (saved !== null) return JSON.parse(saved);
    return window.matchMedia('(prefers-color-scheme: dark)').matches;
  });

  useEffect(() => {
    // Save to localStorage whenever it changes
    localStorage.setItem('darkMode', JSON.stringify(isDark));
    
    // Apply dark mode class to html element
    if (isDark) {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
  }, [isDark]);

  useEffect(() => {
    // Listen for system preference changes (optional)
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    const handleChange = (e) => {
      // Only update if user hasn't set a preference manually
      const saved = localStorage.getItem('darkMode');
      if (saved === null) {
        setIsDark(e.matches);
      }
    };
    
    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  const toggleDarkMode = () => {
    setIsDark(prev => !prev);
  };

  const value = {
    isDark,
    toggleDarkMode,
    setIsDark // Optional: provide direct setter
  };

  return (
    <DarkModeContext.Provider value={value}>
      {children}
    </DarkModeContext.Provider>
  );
};

export { DarkModeContext };