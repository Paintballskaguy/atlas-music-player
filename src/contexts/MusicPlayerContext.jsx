import React, { createContext, useContext, useReducer, useEffect } from 'react';
import ApiService from '../services/api';

const MusicPlayerContext = createContext();

// Action types
const ACTIONS = {
  SET_LOADING: 'SET_LOADING',
  SET_PLAYLIST: 'SET_PLAYLIST',
  SET_CURRENT_SONG: 'SET_CURRENT_SONG',
  SET_CURRENT_SONG_DETAILS: 'SET_CURRENT_SONG_DETAILS',
  SET_LYRICS: 'SET_LYRICS',
  SET_PLAYING: 'SET_PLAYING',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_DURATION: 'SET_DURATION',
  SET_VOLUME: 'SET_VOLUME',
  SET_PLAYBACK_SPEED: 'SET_PLAYBACK_SPEED',
  SET_SHUFFLE: 'SET_SHUFFLE',
  SET_REPEAT: 'SET_REPEAT',
  SET_ERROR: 'SET_ERROR',
  CLEAR_ERROR: 'CLEAR_ERROR'
};

const initialState = {
  playlist: [],
  currentSong: null,
  currentSongDetails: null,
  lyrics: null,
  
  // Playback state
  isPlaying: false,
  currentTime: 0,
  duration: 0,
  volume: 50,
  playbackSpeed: 1,
  shuffle: false,
  repeat: false,
  
  // UI state
  isLoading: true,
  error: null
};

// Reducer
function musicPlayerReducer(state, action) {
  switch (action.type) {
    case ACTIONS.SET_LOADING:
      return { ...state, isLoading: action.payload };
    
    case ACTIONS.SET_PLAYLIST:
      return { ...state, playlist: action.payload };
    
    case ACTIONS.SET_CURRENT_SONG:
      return { ...state, currentSong: action.payload };
    
    case ACTIONS.SET_CURRENT_SONG_DETAILS:
      return { ...state, currentSongDetails: action.payload };
    
    case ACTIONS.SET_LYRICS:
      return { ...state, lyrics: action.payload };
    
    case ACTIONS.SET_PLAYING:
      return { ...state, isPlaying: action.payload };
    
    case ACTIONS.SET_CURRENT_TIME:
      return { ...state, currentTime: action.payload };
    
    case ACTIONS.SET_DURATION:
      return { ...state, duration: action.payload };
    
    case ACTIONS.SET_VOLUME:
      return { ...state, volume: action.payload };
    
    case ACTIONS.SET_PLAYBACK_SPEED:
      return { ...state, playbackSpeed: action.payload };
    
    case ACTIONS.SET_SHUFFLE:
      return { ...state, shuffle: action.payload };
    
    case ACTIONS.SET_REPEAT:
      return { ...state, repeat: action.payload };
    
    case ACTIONS.SET_ERROR:
      return { ...state, error: action.payload, isLoading: false };
    
    case ACTIONS.CLEAR_ERROR:
      return { ...state, error: null };
    
    default:
      return state;
  }
}

// Provider component
export const MusicPlayerProvider = ({ children }) => {
  const [state, dispatch] = useReducer(musicPlayerReducer, initialState);

  // Load initial playlist
  useEffect(() => {
    const loadPlaylist = async () => {
      try {
        dispatch({ type: ACTIONS.SET_LOADING, payload: true });
        const playlist = await ApiService.fetchPlaylist();
        dispatch({ type: ACTIONS.SET_PLAYLIST, payload: playlist });
        
        // Set first song as current if playlist exists
        if (playlist && playlist.length > 0) {
          dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: playlist[0] });
          await loadSongDetails(playlist[0].id);
        }
        
        dispatch({ type: ACTIONS.SET_LOADING, payload: false });
      } catch (error) {
        dispatch({ type: ACTIONS.SET_ERROR, payload: error.message });
      }
    };

    loadPlaylist();
  }, []);

  // Actions
  const loadSongDetails = async (songId) => {
    try {
      const songDetails = await ApiService.fetchSong(songId);
      dispatch({ type: ACTIONS.SET_CURRENT_SONG_DETAILS, payload: songDetails });
    } catch (error) {
      console.error('Failed to load song details:', error);
    }
  };

  const loadLyrics = async (songId) => {
    try {
      const lyricsData = await ApiService.fetchLyrics(songId);
      dispatch({ type: ACTIONS.SET_LYRICS, payload: lyricsData.lyrics });
    } catch (error) {
      console.error('Failed to load lyrics:', error);
    }
  };

  const selectSong = async (song) => {
    dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: song });
    dispatch({ type: ACTIONS.SET_PLAYING, payload: false });
    dispatch({ type: ACTIONS.SET_CURRENT_TIME, payload: 0 });
    await loadSongDetails(song.id);
    await loadLyrics(song.id);
  };

  const togglePlayPause = () => {
    dispatch({ type: ACTIONS.SET_PLAYING, payload: !state.isPlaying });
  };

  const nextSong = () => {
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong?.id);
    if (currentIndex < state.playlist.length - 1) {
      const nextSong = state.playlist[currentIndex + 1];
      selectSong(nextSong);
    }
  };

  const previousSong = () => {
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong?.id);
    if (currentIndex > 0) {
      const prevSong = state.playlist[currentIndex - 1];
      selectSong(prevSong);
    }
  };

  const setVolume = (volume) => {
    dispatch({ type: ACTIONS.SET_VOLUME, payload: volume });
  };

  const setPlaybackSpeed = (speed) => {
    dispatch({ type: ACTIONS.SET_PLAYBACK_SPEED, payload: speed });
  };

  const toggleShuffle = () => {
    dispatch({ type: ACTIONS.SET_SHUFFLE, payload: !state.shuffle });
  };

  const toggleRepeat = () => {
    const repeatModes = [false, 'one', 'all'];
    const currentIndex = repeatModes.indexOf(state.repeat);
    const nextRepeat = repeatModes[(currentIndex + 1) % repeatModes.length];
    dispatch({ type: ACTIONS.SET_REPEAT, payload: nextRepeat });
  };

  const contextValue = {
    ...state,
    // Actions
    selectSong,
    togglePlayPause,
    nextSong,
    previousSong,
    setVolume,
    setPlaybackSpeed,
    toggleShuffle,
    toggleRepeat,
    loadLyrics,
    clearError: () => dispatch({ type: ACTIONS.CLEAR_ERROR })
  };

  return (
    <MusicPlayerContext.Provider value={contextValue}>
      {children}
    </MusicPlayerContext.Provider>
  );
};

// Custom hook
export const useMusicPlayer = () => {
  const context = useContext(MusicPlayerContext);
  if (!context) {
    throw new Error('useMusicPlayer must be used within a MusicPlayerProvider');
  }
  return context;
};