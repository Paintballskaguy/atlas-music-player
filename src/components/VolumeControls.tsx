// src/components/VolumeControls.tsx
import React from 'react';
import { Volume2, VolumeX } from 'lucide-react';
import { useMusicPlayer } from '../contexts/MusicPlayerContext';

const VolumeControls: React.FC = () => {
  const { volume, setVolume } = useMusicPlayer();

  const handleVolumeChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setVolume(parseInt(e.target.value));
  };

  const toggleMute = (): void => {
    setVolume(volume > 0 ? 0 : 50);
  };

  const sliderStyle: React.CSSProperties = {
    background: `linear-gradient(to right, #14b8a6 0%, #14b8a6 ${volume}%, #e5e7eb ${volume}%, #e5e7eb 100%)`
  };

  const darkSliderStyle: React.CSSProperties = {
    background: `linear-gradient(to right, #5eead4 0%, #5eead4 ${volume}%, #4b5563 ${volume}%, #4b5563 100%)`
  };

  return (
    <>
      <div className="flex items-center space-x-4 w-full">
        <button 
          onClick={toggleMute}
          className="p-2 rounded-full bg-custom-yellow-100 dark:bg-custom-yellow-800 hover:bg-custom-yellow-200 dark:hover:bg-custom-yellow-700 transition-all duration-200 hover:scale-110"
          aria-label={volume > 0 ? 'Mute' : 'Unmute'}
        >
          {volume > 0 ? (
            <Volume2 
              size={24} 
              className="text-custom-yellow-700 dark:text-custom-yellow-200" 
            />
          ) : (
            <VolumeX 
              size={24} 
              className="text-custom-red-600 dark:text-custom-red-400" 
            />
          )}
        </button>
        
        <div className="flex-1 relative">
          <input 
            type="range" 
            min="0" 
            max="100" 
            value={volume}
            onChange={handleVolumeChange}
            className="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 dark:hidden"
            style={sliderStyle}
            aria-label="Volume"
          />
          <input 
            type="range" 
            min="0" 
            max="100" 
            value={volume}
            onChange={handleVolumeChange}
            className="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 hidden dark:block"
            style={darkSliderStyle}
            aria-label="Volume"
          />
        </div>
        
        <span className="text-sm font-bold text-custom-teal-600 dark:text-custom-teal-300 min-w-[3rem] text-right">
          {volume}%
        </span>
      </div>
      
      <style dangerouslySetInnerHTML={{
        __html: `
          .volume-slider::-webkit-slider-thumb {
            appearance: none;
            width: 22px;
            height: 22px;
            border-radius: 50%;
            background: #dc2626;
            cursor: pointer;
            border: 3px solid white;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
            transition: all 0.2s ease;
          }
          
          .volume-slider::-webkit-slider-thumb:hover {
            background: #b91c1c;
            transform: scale(1.1);
            box-shadow: 0 6px 12px rgba(0, 0, 0, 0.4);
          }
          
          .volume-slider::-moz-range-thumb {
            width: 22px;
            height: 22px;
            border-radius: 50%;
            background: #dc2626;
            cursor: pointer;
            border: 3px solid white;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
            transition: all 0.2s ease;
          }
          
          .volume-slider::-moz-range-thumb:hover {
            background: #b91c1c;
            transform: scale(1.1);
            box-shadow: 0 6px 12px rgba(0, 0, 0, 0.4);
          }
        `
      }} />
    </>
  );
};

export default VolumeControls;