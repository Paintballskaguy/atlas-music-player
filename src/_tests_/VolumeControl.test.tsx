import { describe, it, expect, vi, beforeEach } from 'vitest'
import { render, screen, fireEvent } from '@testing-library/react'
import VolumeControls from '../components/VolumeControls'
// FIX 1: Change import from the raw Context object to the custom hook
import { useMusicPlayer } from '../contexts/MusicPlayerContext' 
import React from 'react'

// Mock Lucide React icons
vi.mock('lucide-react', () => ({
  Volume2: ({ size, className }: any) => <div data-testid="volume-icon" data-type="volume2" className={className}>Volume2Icon</div>,
  VolumeX: ({ size, className }: any) => <div data-testid="volume-icon" data-type="volumex" className={className}>VolumeXIcon</div>,
}))

// FIX 2: Mock the useMusicPlayer hook to control the values returned to the component
vi.mock('../contexts/MusicPlayerContext', () => ({
  useMusicPlayer: vi.fn(),
}))

describe('VolumeControls', () => {
  const mockSetVolume = vi.fn()
  
  const defaultContextValue = {
    volume: 50,
    setVolume: mockSetVolume,
    // Add other required context properties
    currentSongDetails: null,
    lyrics: null,
    currentSong: null,
    loadLyrics: vi.fn(),
    playSong: vi.fn(),
    pauseSong: vi.fn(),
    nextSong: vi.fn(),
    previousSong: vi.fn(),
    isPlaying: false,
  }

  // FIX 3: Update the helper function to set the mock return value and render the component directly
  const renderWithContext = (contextValue = defaultContextValue) => {
    // Set the mock implementation for the hook
    vi.mocked(useMusicPlayer).mockReturnValue(contextValue)
    
    // Render the component which will now use the mocked hook result
    return render(<VolumeControls />)
  }

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('renders with default volume (50%)', () => {
    const { container } = renderWithContext()
    
    expect(screen.getByText('50%')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with volume at 0% (muted)', () => {
    const contextValue = {
      ...defaultContextValue,
      volume: 0
    }
    
    const { container } = renderWithContext(contextValue)
    
    expect(screen.getByText('0%')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with volume at 100%', () => {
    const contextValue = {
      ...defaultContextValue,
      volume: 100
    }
    
    const { container } = renderWithContext(contextValue)
    
    expect(screen.getByText('100%')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls setVolume when slider is changed', () => {
    renderWithContext()
    
    const sliders = screen.getAllByLabelText('Volume')
    const lightModeSlider = sliders[0] // First slider is for light mode
    
    fireEvent.change(lightModeSlider, { target: { value: '75' } })
    
    expect(mockSetVolume).toHaveBeenCalledWith(75)
  })

  it('toggles mute/unmute when button is clicked', () => {
    renderWithContext()
    
    const muteButton = screen.getByRole('button', { name: /mute/i })
    fireEvent.click(muteButton)
    
    expect(mockSetVolume).toHaveBeenCalledWith(0)
  })

  it('unmutes when button is clicked and volume is 0', () => {
    const contextValue = {
      ...defaultContextValue,
      volume: 0
    }
    
    renderWithContext(contextValue)
    
    const unmuteButton = screen.getByRole('button', { name: /unmute/i })
    fireEvent.click(unmuteButton)
    
    expect(mockSetVolume).toHaveBeenCalledWith(50)
  })

  it('shows VolumeX icon when volume is 0', () => {
    const contextValue = {
      ...defaultContextValue,
      volume: 0
    }
    
    renderWithContext(contextValue)
    
    const volumeIcon = screen.getByTestId('volume-icon')
    expect(volumeIcon).toHaveAttribute('data-type', 'volumex')
    expect(volumeIcon).toHaveClass('text-custom-red-600')
  })

  it('shows Volume2 icon when volume is greater than 0', () => {
    renderWithContext()
    
    const volumeIcon = screen.getByTestId('volume-icon')
    expect(volumeIcon).toHaveAttribute('data-type', 'volume2')
    expect(volumeIcon).toHaveClass('text-custom-yellow-700')
  })

  it('applies correct slider styles based on volume', () => {
    const contextValue = {
      ...defaultContextValue,
      volume: 75
    }
    
    renderWithContext(contextValue)
    
    const sliders = screen.getAllByLabelText('Volume')
    const lightModeSlider = sliders[0]
    
    // Check that the style includes the gradient with 75%
    expect(lightModeSlider.style.background).toContain('14b8a6 0%')
    expect(lightModeSlider.style.background).toContain('14b8a6 75%')
  })
})