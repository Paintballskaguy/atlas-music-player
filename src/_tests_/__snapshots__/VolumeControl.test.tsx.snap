// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VolumeControls > renders with default volume (50%) 1`] = `
<div
  class="flex items-center space-x-4 w-full"
>
  <button
    aria-label="Mute"
    class="p-2 rounded-full bg-custom-yellow-100 dark:bg-custom-yellow-800 hover:bg-custom-yellow-200 dark:hover:bg-custom-yellow-700 transition-all duration-200 hover:scale-110"
  >
    <div
      class="text-custom-yellow-700 dark:text-custom-yellow-200"
      data-testid="volume-icon"
      data-type="volume2"
    >
      Volume2Icon
    </div>
  </button>
  <div
    class="flex-1 relative"
  >
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 dark:hidden"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #14b8a6 0%, #14b8a6 50%, #e5e7eb 50%, #e5e7eb 100%);"
      type="range"
      value="50"
    />
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 hidden dark:block"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #5eead4 0%, #5eead4 50%, #4b5563 50%, #4b5563 100%);"
      type="range"
      value="50"
    />
  </div>
  <span
    class="text-sm font-bold text-custom-teal-600 dark:text-custom-teal-300 min-w-[3rem] text-right"
  >
    50
    %
  </span>
</div>
`;

exports[`VolumeControls > renders with volume at 0% (muted) 1`] = `
<div
  class="flex items-center space-x-4 w-full"
>
  <button
    aria-label="Unmute"
    class="p-2 rounded-full bg-custom-yellow-100 dark:bg-custom-yellow-800 hover:bg-custom-yellow-200 dark:hover:bg-custom-yellow-700 transition-all duration-200 hover:scale-110"
  >
    <div
      class="text-custom-red-600 dark:text-custom-red-400"
      data-testid="volume-icon"
      data-type="volumex"
    >
      VolumeXIcon
    </div>
  </button>
  <div
    class="flex-1 relative"
  >
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 dark:hidden"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #14b8a6 0%, #14b8a6 0%, #e5e7eb 0%, #e5e7eb 100%);"
      type="range"
      value="0"
    />
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 hidden dark:block"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #5eead4 0%, #5eead4 0%, #4b5563 0%, #4b5563 100%);"
      type="range"
      value="0"
    />
  </div>
  <span
    class="text-sm font-bold text-custom-teal-600 dark:text-custom-teal-300 min-w-[3rem] text-right"
  >
    0
    %
  </span>
</div>
`;

exports[`VolumeControls > renders with volume at 100% 1`] = `
<div
  class="flex items-center space-x-4 w-full"
>
  <button
    aria-label="Mute"
    class="p-2 rounded-full bg-custom-yellow-100 dark:bg-custom-yellow-800 hover:bg-custom-yellow-200 dark:hover:bg-custom-yellow-700 transition-all duration-200 hover:scale-110"
  >
    <div
      class="text-custom-yellow-700 dark:text-custom-yellow-200"
      data-testid="volume-icon"
      data-type="volume2"
    >
      Volume2Icon
    </div>
  </button>
  <div
    class="flex-1 relative"
  >
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 dark:hidden"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #14b8a6 0%, #14b8a6 100%, #e5e7eb 100%, #e5e7eb 100%);"
      type="range"
      value="100"
    />
    <input
      aria-label="Volume"
      class="volume-slider w-full h-3 rounded-full appearance-none cursor-pointer transition-all duration-200 hover:scale-y-110 hidden dark:block"
      max="100"
      min="0"
      style="background: linear-gradient(to right, #5eead4 0%, #5eead4 100%, #4b5563 100%, #4b5563 100%);"
      type="range"
      value="100"
    />
  </div>
  <span
    class="text-sm font-bold text-custom-teal-600 dark:text-custom-teal-300 min-w-[3rem] text-right"
  >
    100
    %
  </span>
</div>
`;
